File: research/tcp/tcp.client.cpp
/**
g++ tcp.client.cpp -g -O0 -o tcp.client && ./tcp.client 1990 4096 
*/
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <netinet/in.h>

#define srs_trace(msg, ...)   printf(msg, ##__VA_ARGS__);printf("\n")

int main(int argc, char** argv)
{
    srs_trace("tcp client to recv bytes from server");
    if (argc <= 2) {
        srs_trace("Usage: %s <port> <packet_bytes>", argv[0]);
        srs_trace("   port: the port to connect to.");
        srs_trace("   packet_bytes: the bytes for packet to send.");
        srs_trace("For example:");
        srs_trace("   %s %d %d", argv[0], 1990, 4096);
        return -1;
    }
    
    int server_port = ::atoi(argv[1]);
    int packet_bytes = ::atoi(argv[2]);
    srs_trace("server_port is %d", server_port);
    srs_trace("packet_bytes is %d", packet_bytes);
    
    int fd = socket(AF_INET, SOCK_STREAM, 0);
    if (fd < 0) {
        srs_trace("create socket failed.");
        return -1;
    }
    
    sockaddr_in addr;
    addr.sin_family = AF_INET;
    addr.sin_port = htons(server_port);
    addr.sin_addr.s_addr = inet_addr("127.0.0.1");
    if (::connect(fd, (sockaddr*)&addr, sizeof(addr)) == -1) {
        srs_trace("connect server error.");
        return -1;
    }
    srs_trace("connect server success. fd=%d", fd);
    
    // get the sockoptions
    int sock_recv_buffer = 0;
    socklen_t nb_sock_recv_buffer = sizeof(int);
    if (getsockopt(fd, SOL_SOCKET, SO_RCVBUF, &sock_recv_buffer, &nb_sock_recv_buffer) < 0) {
        srs_trace("get sockopt failed.");
        return -1;
    }
    srs_trace("SO_RCVBUF=%d", sock_recv_buffer);
    
    char b[4096];
    for (;;) {
        ssize_t nb_recv = recv(fd, b, sizeof(b), 0);
        if (nb_recv <= 0) {
            srs_trace("recv bytes to socket error.");
            ::close(fd);
            break;
        }
    }
    srs_trace("completed");
    
    return 0;
}

File: research/tcp/tcp.client.py
'''
python tcp.client.py 127.0.0.1 1990 4096
'''
import socket, sys
if len(sys.argv) <= 3:
    print("Usage: %s <ip> <port> <packet_bytes>"%(sys.argv[0]))
    print("   ip: the ip to connect to.")
    print("   port: the port to connect to.")
    print("   packet_bytes: the bytes for packet to send.")
    print("For example:")
    print("   %s %d %d"%(sys.argv[0], "127.0.0.1", 1990, 4096))
    sys.exit(-1)

server_ip = str(sys.argv[1])
listen_port = int(sys.argv[2])
packet_bytes = int(sys.argv[3])
print("server_ip is %s"%server_ip)
print("listen_port is %d"%listen_port)
print("packet_bytes is %d"%packet_bytes)

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

s.connect((server_ip, listen_port))
print("connect socket %s:%d success."%(server_ip, listen_port))
    
while True:
    data = s.recv(1024)
    if not data: 
        break
s.close()
